crucible-go/src/Lang/Crucible/Go/Simulate.hs

```diff
@@ -73,20 +73,20 @@ setSimulatorVerbosity verbosity sym = do
   return ()
 
 goExtensionEval ::
-  forall sym p ext rtp blocks r ctx.
-  (IsSymInterface sym) =>
-  sym ->
+  forall sym bak p ext rtp blocks r ctx.
+  (IsSymInterface sym, IsBoolSolver sym bak) =>
```

Comment:
It's a good question, and I don't know if I have a firm opinion.  For the initial refactoring, it was easier to have them be separate, I think, to help the type checker guide me to places where changes needed to happen. Now, I'm not sure what the best way to arrange this is.  

I'm also not necessarily a huge fan of the names of these constraints.  Any ideas?

Reply:
I'm generally fond of the `IsSymInterface` approach of bundling all of the common classes up into a type synonym simply because it makes it easier to write code for common cases. This is especially valuable when I write applications on top of `crucible`, since I don't have to think about the various distinctions between each class's expressive power, decreasing my overall cognitive burden.

I can't think of many scenarios where you'd shoot yourself in the foot by picking such a class (is there such a scenario?), so I'd favor offering one. I'm not great at coming up with names, but perhaps `IsSymSolverInterface` would do. I don't think you need to go through the effort of using it everywhere in `crucible`, however, since it's ultimately a convenience for users.

> I'm also not necessarily a huge fan of the names of these constraints. Any ideas?

I struggle to think of better names, primarily since these classes represent a grab-bag collection of ways to interact with expressions and solvers in various ways. I suppose you could object to the use of "`Bool`" in `IsBoolSolver`, but I'm not convinced that a new name would make its purpose more obvious. For better or worse, I think you just have to read the Haddocks to figure out what it's about.

Reply:
Thinking about this some more, I think the issue is that there are a lot of code paths where the `bak` type variable is hidden, so it makes it hard to carry around a single combined constraint that mentions both.  That leaves you with a few options 1) two flavors (one that mentions `bak` and one that doesn't) 2) two separate constraints, as here or 3) make `IsBoolSolver` take the constituents of `IsSymInterface` as superclass constraints.

Maybe 3 is best?

Reply:
I'd be happy with option (3). The Haddocks for `IsBoolSolver sym bak` even state that `sym` is expected to satisfy the `IsSymInterface` constraints, but this is currently not enforced, which seems odd.